File: kotlin/src/main/kotlin/com/cardinal/modules/ShareHcp.kt
package com.cardinal.modules

import com.cardinal.sdk.createSdk
import com.icure.cardinal.sdk.model.DecryptedDocument
import com.icure.cardinal.sdk.model.embed.AccessLevel
import com.cardinal.utils.prettyPrint
import com.icure.cardinal.sdk.CardinalSdk
import java.util.*

suspend fun shareWithHcp(sdk: CardinalSdk) {
	try {
		print("Login of the other hcp: ")
		val username = readln().trim()
		print("Insert the password for this hcp: ")
		val otherPassword = readln()
		val otherSdk = createSdk(username, otherPassword)
		val otherHcp = otherSdk.healthcareParty.getCurrentHealthcareParty()

		val oldDocument = sdk.document.createDocument(
			DecryptedDocument(
				id = UUID.randomUUID().toString(),
				name = "An important document"
			).let {
				sdk.document.withEncryptionMetadata(it, null)
			}
		)
		prettyPrint(oldDocument)

		try {
			otherSdk.document.getDocument(oldDocument.id)
		} catch (e: Exception) {
			println("This means I am not authorized to read the document -> ${e.message}")
		}

		sdk.document.shareWith(
			delegateId = otherHcp.id,
			document = oldDocument
		)

		val oldDocumentOtherHcp = otherSdk.document.getDocument(oldDocument.id)
		prettyPrint(oldDocumentOtherHcp)

		val newDocument = DecryptedDocument(
			id = UUID.randomUUID().toString(),
			name = "Another important document"
		)
		val newDocumentWithMetadata = sdk.document.withEncryptionMetadata(
			newDocument,
			null,
			delegates = mapOf(otherHcp.id to AccessLevel.Read)
		)
		val createdNewDocument = sdk.document.createDocument(newDocumentWithMetadata)

		val newDocumentOtherHcp = otherSdk.document.getDocument(createdNewDocument.id)
		prettyPrint(newDocumentOtherHcp)

	} catch (e: Exception) {
		println("Something went wrong: ${e.message}")
	}
}

File: python/src/modules/share_hcp.py
import uuid
from create_sdk import create_cardinal_sdk
from cardinal_sdk import CardinalSdk
from cardinal_sdk.model import DecryptedDocument, SimpleShareResultDecryptedDocumentSuccess, AccessLevel
from utils import pretty_print_document


def share_with_hcp(sdk: CardinalSdk):
	try:
		username = input("Login of the other hcp: ").strip()
		other_password = input("Insert the password for this hcp: ")
		other_sdk = create_cardinal_sdk(username, other_password)
		other_hcp = other_sdk.healthcare_party.get_current_healthcare_party_blocking()

		old_document_without_encryption_meta = DecryptedDocument(
			id=str(uuid.uuid4()),
			name="An important document"
		)
		old_document = sdk.document.create_document_blocking(
			sdk.document.with_encryption_metadata_blocking(old_document_without_encryption_meta, None)
		)
		pretty_print_document(old_document)

		try:
			other_sdk.document.get_document_blocking(old_document.id)
		except Exception as e:
			print(f"This means I am not authorized to read the document -> {e}")

		result = sdk.document.share_with_blocking(
			delegate_id=other_hcp.id,
			document=old_document
		)

		if isinstance(result, SimpleShareResultDecryptedDocumentSuccess):
			print("Successfully shared document")

		old_document_other_hcp = other_sdk.document.get_document_blocking(old_document.id)
		pretty_print_document(old_document_other_hcp)

		new_document = DecryptedDocument(
			id=str(uuid.uuid4()),
			name="Another important document"
		)
		new_document_with_metadata = sdk.document.with_encryption_metadata_blocking(
			new_document,
			None,
			delegates={other_hcp.id: AccessLevel.Read}
		)
		created_new_document = sdk.document.create_document_blocking(new_document_with_metadata)

		new_document_other_hcp = other_sdk.document.get_document_blocking(created_new_document.id)
		pretty_print_document(new_document_other_hcp)
	except Exception as e:
		print(f"Something went wrong: {e}")
